commons/zenoh-util/src/log.rs

```diff
@@ -11,86 +11,260 @@
+///
+/// # Use in tests
+///
+/// This function should **not** be used in tests, as it would panic as soon as there is more
+/// than one test executed in the same unit, because only the first one to execute would be able to
+/// register the global tracing subscriber.
+///
+/// Moreover, `tracing` and Rust logging in general requires special care about testing because of
+/// libtest output capturing; see
+/// [`SubscriberBuilder::with_test_writer`](tracing_subscriber::fmt::SubscriberBuilder::with_test_writer).
+///
+/// That's why we advise you to use a dedicated library like [`test-log`][3]
+/// (with `"tracing"` feature enabled).
+///
+/// # Memory leak
+///
+/// [`tracing`] use a static subscriber, which is not deallocated prior to process exiting.
+/// Tools such as `valgrind` will then report memory leaks in *still reachable* category.
+///
+/// However, when `RUST_LOG` is provided (see [Advanced use](#advanced-use)),
+/// [`tracing_subscriber::EnvFilter`] is then used, and causes also *still reachable* blocks,
+/// but also *possibly lost* blocks, which are [known false-positives][2].
+/// Those "leaks" can be "suppressed" from `valgrind` report using the following suppression:
+/// ```text
+/// {
+///    zenoh_init_logging
+///    Memcheck:Leak
+///    ...
+///    fun:*zenoh*init_logging*
+/// }
+/// ```
+///
+/// [1]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives
+/// [2]: https://github.com/rust-lang/regex/issues/1205
+/// [3]: https://crates.io/crates/test-log
+pub fn init_logging() {
+    try_init_logging().expect(ALREADY_INITIALIZED);
 }
 
-fn init_env_filter(env_filter: EnvFilter) {
-    let subscriber = tracing_subscriber::fmt()
-        .with_env_filter(env_filter)
+/// Initialize zenoh logging using the provided logging level.
+///
+/// See [`init_logging`] if you prefer a dynamic setting the level using an environment variable.
+/// Logs are printed on stdout and are formatted like the following:
+/// ```text
+/// 2024-06-19T09:46:18.808602Z  INFO main ThreadId(01) zenoh::net::runtime: Using ZID: 1a615ea88fe1dc531a9d8701775d5bee
+/// 2024-06-19T09:46:18.814577Z  INFO main ThreadId(01) zenoh::net::runtime::orchestrator: Zenoh can be reached at: tcp/[fe80::1]:58977
+/// ```
+///
+/// # Advanced use
+///
+/// zenoh logging uses `tracing` crates internally; this function is just a convenient wrapper
+/// around `tracing-subscriber`. If you want to control the formatting, or have a finer grain on
+/// log filtering, we advise using `tracing-subscriber` directly.
+///
+/// However, to make migration and on the fly debugging easier, [`RUST_LOG`][1] environment variable
+/// can still be used, and will override the provided level.
+///
+/// # Panics
+///
+/// This function may panic if the following operations fail:
+/// - parsing `RUST_LOG` (see [Advanced use](#advanced-use)) environment variable
```

Comment:
parsing `RUST_LOG` / `ZENOH_LOG`

Reply:
I didn't get this, sorry. But why? Isn't it just adds more complexity?

Reply:
We talked about it in our last meeting: to keep things the simplest possible, I've dissociated initialization from variable and initialization from code.
It's still possible to override the provided level with `RUST_LOG` environment variable for debug purpose.

I don't know what would be the use case of using `init_logging_from_level` together with `ZENOH_LOG`. The only thing I see is when you want to increase the logging level to debug something (without having to recompile/redeploy), but in that case, `RUST_LOG` seems more appropriate to me, and it's already supported.